### DISCUSSION TITLE
Something doesn't work correctly

### DISCUSSION THREAD

**Discussion by MahdisYavari** [skydoves/Balloon#252] [Q&A] [answered]
Actually I set the balloon width and height "wrap" but it doesn't work for me and I tried a lot .
.setWidth(BalloonSizeSpec.WRAP)
.setHeight(BalloonSizeSpec.WRAP)

**Comment by skydoves** [comment]
Hi @MahdisYavari ,
Could you share your custom layout and `Balloon.Builder` relative codes?

**Reply by MahdisYavari** [reply]
Yes of course .
This is my balloon.builder class .

fun getEditBalloon(context: Context, lifecycleOwner: LifecycleOwner): Balloon {
    return Balloon.Builder(context)
      .setText("payment detail")
      .setArrowSize(10)
      .setWidthRatio(1.0f)
      .setHeight(BalloonSizeSpec.WRAP)
      .setWidth(BalloonSizeSpec.WRAP)
      .setArrowPosition(0.5f)
      .setArrowSize(10)
      .setPadding(12)
      .setMarginRight(12)
      .setMarginLeft(12)
      .setTextSize(15f)
      .setCornerRadius(8f)
      .setTextColorResource(R.color.white)
      .setDismissWhenClicked(true)
      .setDismissWhenTouchOutside(false)
      .setBackgroundColorResource(R.color.darkGray)
      .setArrowPositionRules(ArrowPositionRules.ALIGN_ANCHOR)
      .setArrowOrientation(ArrowOrientation.BOTTOM)
      .setBalloonAnimation(BalloonAnimation.ELASTIC)
      .setLifecycleOwner(lifecycleOwner)
      .build()
  }

**Reply by MahdisYavari** [reply]
I also try this parameter " setLayout " to my balloon builder and it didn't work. I just want to show a text inside the tooltip but it's "match parent" .

this is a piece of my simple layout code : 

<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:app="http://schemas.android.com/apk/res-auto"
    android:layout_width="wrap_content"
    android:layout_height="wrap_content"
    xmlns:tools="http://schemas.android.com/tools">

    <com.google.android.material.textview.MaterialTextView
        android:id="@+id/txt_balloon_content"
        android:layout_width="wrap_content"
        android:layout_height="wrap_content"
        app:layout_constraintBottom_toBottomOf="parent"
        app:layout_constraintEnd_toEndOf="parent"
        app:layout_constraintStart_toStartOf="parent"
        app:layout_constraintTop_toTopOf="parent"
        android:text="payment detail" />

</androidx.constraintlayout.widget.ConstraintLayout>

**Reply by MahdisYavari** [reply]
@skydoves please help me , thanks

**Reply by skydoves** [reply]
Apologies for the delayed response.
I recommend using like this for the `MaterialTextView`.

```kotlin
<com.google.android.material.textview.MaterialTextView
    android:id="@+id/txt_balloon_content"
    android:layout_width="wrap_content"
    android:layout_height="wrap_content"
   ..
```

And could you share the screenshot? It will be helpful to track your issue.

**Comment by MahdisYavari** [comment]
![Screenshot (20)](https://user-images.githubusercontent.com/55578318/136188157-ad1391a3-2039-482c-ae25-a64e5c9428c2.png)


This is my balloon tooltip , Actually I changed the root layout to MaterialTextView and I set a red background color to find out its width and height length . Unfortunately I think there is a problem with the whole balloon class and it can't set it to WrapContent . I don't know the reason .

**Reply by MahdisYavari** [reply]
@skydoves

**Reply by skydoves** [reply]
Is the same result if you use the `TextView` or `AppCompatTextView` instead the `MaterialTextView`?

**Reply by skydoves** [reply]
And I the layout style has some little error because the root layout is the `ConstraintLayout `.
The layout_width and layout_height of the `MaterialTextView` should be `0dp`.

```xml
<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
xmlns:app="http://schemas.android.com/apk/res-auto"
android:layout_width="wrap_content"
android:layout_height="wrap_content"
xmlns:tools="http://schemas.android.com/tools">

<com.google.android.material.textview.MaterialTextView
    android:id="@+id/txt_balloon_content"
    android:layout_width="0dp"
    android:layout_height="0dp"
    app:layout_constraintBottom_toBottomOf="parent"
    app:layout_constraintEnd_toEndOf="parent"
    app:layout_constraintStart_toStartOf="parent"
    app:layout_constraintTop_toTopOf="parent"
    android:text="payment detail" />
```

**Reply by MahdisYavari** [reply]
I'm so sorry . I tried all of your recommendations right now . even I set a dp number to its width and height but nothing changed .
The height is changing but the width isn't changed .
@skydoves

**Reply by MahdisYavari** [reply]
Is there any way to fix it ?

**Reply by skydoves** [reply]
No problem. Hmm.. I built the [demo project](https://github.com/skydoves/Balloon/blob/main/app/src/main/res/layout/layout_custom_profile.xml) with little bit updates:

```xml
<?xml version="1.0" encoding="utf-8"?>
<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
  xmlns:app="http://schemas.android.com/apk/res-auto"
  xmlns:tools="http://schemas.android.com/tools"
  android:layout_width="wrap_content"
  android:layout_height="wrap_content"
  android:gravity="center"
  android:orientation="vertical"
  tools:background="@color/background800">

  <androidx.appcompat.widget.AppCompatImageView
    android:id="@+id/icon"
    android:layout_width="wrap_content"
    android:layout_height="wrap_content"
    android:layout_marginStart="8dp"
    android:layout_marginEnd="8dp"
    android:src="@drawable/ic_cards"
    app:layout_constraintBottom_toBottomOf="@+id/name"
    app:layout_constraintEnd_toStartOf="@id/name"
    app:layout_constraintHorizontal_chainStyle="packed"
    app:layout_constraintStart_toStartOf="parent"
    app:layout_constraintTop_toTopOf="@id/name" />

  <TextView
    android:id="@+id/name"
    android:layout_width="wrap_content"
    android:layout_height="wrap_content"
    android:layout_marginTop="8dp"
    android:gravity="center"
    android:includeFontPadding="false"
    android:text="skydoves"
    android:textColor="@color/white"
    android:textSize="16sp"
    android:textStyle="bold"
    app:layout_constraintEnd_toEndOf="parent"
    app:layout_constraintStart_toEndOf="@+id/icon"
    app:layout_constraintTop_toTopOf="parent"
    tools:ignore="HardcodedText" />

  <TextView
    android:id="@+id/content"
    android:layout_width="wrap_content"
    android:layout_height="wrap_content"
    android:layout_marginTop="12dp"
    android:gravity="center"
    android:text="Love coffee, music, magic tricks."
    android:textColor="@color/white"
    android:textSize="14sp"
    app:layout_constraintEnd_toEndOf="parent"
    app:layout_constraintStart_toStartOf="parent"
    app:layout_constraintTop_toBottomOf="@id/name"
    tools:ignore="HardcodedText" />

  <Button
    android:id="@+id/button_edit"
    android:layout_width="match_parent"
    android:layout_height="36dp"
    android:layout_marginTop="16dp"
    android:background="@drawable/round_button_skyblue"
    android:text="Edit"
    android:textAllCaps="false"
    android:textColor="@color/white"
    app:layout_constraintEnd_toEndOf="parent"
    app:layout_constraintStart_toStartOf="parent"
    app:layout_constraintTop_toBottomOf="@id/content"
    tools:ignore="HardcodedText" />

</androidx.constraintlayout.widget.ConstraintLayout>
```

And it works well like the below.

![screenshot1506490675](https://user-images.githubusercontent.com/24237865/136194617-96f1efe7-2eb4-407d-8e2b-3bcba5ee6d39.png)

The demo project may help your problem.

**Reply by MahdisYavari** [reply]
Many thanks for your guidance I will check it again I hope it solve my problem .

**Comment by MahdisYavari** [comment, marked as answer]
@skydoves 
I found the problem . There wasn't any problem with layout file .
 I used this before :
`private val searchBalloon by lazy { BalloonUtils.getEditBalloon(requireContext(), this)` 
but I changed it like the code you wrote :
`private val customProfileBalloon by balloon<BalloonUtils.ProfileBalloonFactory>()`

finally I think it was because of "this" that I set it as lifecycle owner. 
At the end **Thanks** .
